core/common/src/main/java/alluxio/underfs/UnderFileSystem.java

```diff
@@ -88,6 +91,87 @@ public int getValue() {
   }
 
   /**
+   * A class used to cache UnderFileSystem.
+   */
+  private static class Cache {
+    private Map<Key, UnderFileSystem> mUnderFileSystemMap = new ConcurrentHashMap<>();
+
+    Cache() {}
+
+    /**
+     * Get a UFS instance from the cache if exists. Otherwise, creates a new instance and add
+     * that to the cache.
+     *
+     * @param path the UFS path
+     * @param ufsConf the ufs configuration
+     * @param configuration the alluxio configuration
+     * @return the UFS instance
+     */
+    UnderFileSystem get(String path, Object ufsConf, Configuration configuration) {
+      UnderFileSystem fs;
+      Key key = new Key(new AlluxioURI(path));
+      fs = mUnderFileSystemMap.get(key);
+      if (fs != null) {
+        return fs;
+      }
+      fs = UnderFileSystemRegistry.create(path, configuration, ufsConf);
+      // COMPATIBILITY: We need to cast mMap to ConcurrentHashMap to make sure the code can compile
+      // on Java 7 because the Map#putIfAbsent() method has only been introduced in Java 8.
+      UnderFileSystem oldFs =
+          ((ConcurrentHashMap<Key, UnderFileSystem>) mUnderFileSystemMap).putIfAbsent(key, fs);
```

Comment:
Done. I actually copied the code from alluxio/collections/ConcurrentHashSet.java.